ARCs/arc-0200.md

```diff
@@ -0,0 +1,231 @@
+
+# Algorand Smart Contract Token Specification
+
+## Abstract
+
+This ARC (Algorand Request for Comments) specifies an interface for tokens to be implemented on Algorand as smart contracts. The interface defines a minimal interface required for tokens to be held and transferred, with the potential for further augmentation through additional standard interfaces and custom methods.
+
+## Motivation
+
+Currently, most tokens in the Algorand ecosystem are represented by ASAs (Algorand Standard Assets). However, to provide rich extra functionality, it can be desirable to implement tokens as smart contracts instead. To foster an interoperable token ecosystem, it is necessary that the core interfaces for tokens be standardized.
+
+## Specification
+
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.
+
+### Core Token specification
+
+A smart contract token that is compliant with this standard MUST implement the following interface:
+
+```json
+{
+  "name": "ARC-200",
+  "desc": "Smart Contract Token Base Interface",
+  "methods": [
+    {
+      "name": "name",
+      "desc": "Returns the name of the token",
+      "readonly": true,
+      "args": [],
+      "returns": { "type": "bytes[32]", "desc": "The name of the token" }
+    },
+    {
+      "name": "symbol",
+      "desc": "Returns the symbol of the token",
+      "readonly": true,
+      "args": [],
+      "returns": { "type": "bytes[8]", "desc": "The symbol of the token" }
+    },
+    {
+      "name": "decimals",
+      "desc": "Returns the decimals of the token",
+      "readonly": true,
+      "args": [],
+      "returns": { "type": "uint8", "desc": "The decimals of the token" }
+    },
+    {
+      "name": "totalSupply",
+      "desc": "Returns the total supply of the token",
+      "readonly": true,
+      "args": [],
+      "returns": { "type": "uint256", "desc": "The total supply of the token" }
+    },
+    {
+      "name": "balanceOf",
+      "desc": "Returns the current balance of the owner of the token",
+      "readonly": true,
+      "args": [
+        {
+          "type": "address",
+          "name": "owner",
+          "desc": "The address of the owner of the token"
+        }
+      ],
+      "returns": {
+        "type": "uint256",
+        "desc": "The current balance of the holder of the token"
+      }
+    },
+    {
+      "name": "transfer",
+      "desc": "Transfers tokens",
+      "readonly": false,
+      "args": [
+        {
+          "type": "address",
+          "name": "to",
+          "desc": "The destination of the transfer"
+        },
+        {
+          "type": "uint256",
+          "name": "value",
+          "desc": "Amount of tokens to transfer"
+        }
+      ],
+      "returns": { "type": "bool", "desc": "Success" }
```

Comment:
Do you see any advantage in making `transfer` and `transferFrom` return void instead of booleans? My impression is that for ERC-20 implementations, the safer thing to do is to only ever return true or error if the operation isn't possible. Otherwise it's very important that callers check the returned value, which is just another opportunity for a bug.

Reply:
It does trivially return True. To be on the safe side you would check that it does despite the fact that in practice the only alternative is to throw an error. 
I think there is no advantage of returning void instead of boolean. Same problem more uncertainty.
I know we are not going into implementation but if we were I would advise having both methods return true if not error and don't even bother having code returning false.

Reply:
In other words, it is not written in the spec but these methods only return true, which is a bool

Reply:
Then why not write into the spec that the methods should return only true? Or better yet, remove all uncertainty and have it return void?

True, this would be a departure from the ERC-20 spec, but a very small one that's easily bridged.

Reply:
I think the argument for `bool` is that one can attempt to do a transfer which is a feature that isn't otherwise possible. The counterargument is that one should just check `allowance` before attempting a transfer.

Reply:
I think bool is fine the way it is. I don't think we should bake into the spec the fact that on the underlying network if something does not match up, assertions fail, it will crash and burn.